feos-dft/src/interface/mod.rs

```diff
@@ -196,6 +196,142 @@ impl<U: EosUnit, F: HelmholtzEnergyFunctional> PlanarInterface<U, F> {
         self
     }
 
+    /// Relative adsorption of component `i' with respect to `j': \Gamma_i^(j)
+    pub fn relative_adsorption(&self) -> EosResult<QuantityArray2<U>> {
+        let s = self.profile.density.shape();
+        let mut rho_l = Array1::zeros(s[0]) * U::reference_density();
+        let mut rho_v = Array1::zeros(s[0]) * U::reference_density();
+
+        // Calculate the partial densities in the liquid and in the vapor phase
+        for i in 0..s[0] {
+            // rho_l.try_set(i, self.profile.density.get((i, 0)) * m[i])?;
+            // rho_v.try_set(i, self.profile.density.get((i, s[1] - 1)) * m[i])?;
+            rho_l.try_set(i, self.profile.density.get((i, 0)))?;
+            rho_v.try_set(i, self.profile.density.get((i, s[1] - 1)))?;
+        }
+
+        // Calculate \Gamma_i^(j)
+        let gamma = Array2::from_shape_fn((s[0], s[0]), |(i, j)| -> f64 {
+            if i == j {
+                0.0
+            } else {
+                (-(rho_l.get(i) - rho_v.get(i))
+                    * ((&self.profile.density.index_axis(Axis_nd(0), j) - rho_l.get(j))
+                        / (rho_l.get(j) - rho_v.get(j))
+                        - (&self.profile.density.index_axis(Axis_nd(0), i) - rho_l.get(i))
+                            / (rho_l.get(i) - rho_v.get(i)))
+                    .integrate(&self.profile.grid.integration_weights_unit()))
+                .to_reduced(U::reference_density() * U::reference_length())

```

Comment:
There is no need to used reduced units here. You can use `QuantityArray2::from_shape_fn`

Reply:
Thanks, that simplifies it a bit.